Mathlib/Algebra/Category/ModuleCat/Presheaf.lean

```diff
@@ -203,4 +234,194 @@ instance : Faithful (toPresheaf R) where
+noncomputable def restriction {X Y : Cᵒᵖ} (f : X ⟶ Y) :
+    evaluation R X ⟶ evaluation R Y ⋙ ModuleCat.restrictScalars (R.map f) where
+  app := restrictionApp f
+  naturality := fun M N φ => by
+    ext x
+    exact (congr_hom (φ.hom.naturality f) x).symm
+
+variable {R}
+
+attribute [local simp] restrictionApp_apply
+
+lemma restriction_app_id (M : PresheafOfModules R) (X : Cᵒᵖ) :
+    restrictionApp (𝟙 X) M =
+      (ModuleCat.restrictScalarsId' (R.map (𝟙 X)) (R.map_id X)).inv.app (M.obj X) := by aesop
+
+lemma restriction_app_comp (M : PresheafOfModules R) {X Y Z : Cᵒᵖ} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    restrictionApp (f ≫ g) M =
+      restrictionApp f M ≫
+        (ModuleCat.restrictScalars (R.map f)).map (restrictionApp g M) ≫
+        (ModuleCat.restrictScalarsComp' _ _ _ (R.map_comp f g)).inv.app (M.obj Z) := by aesop
+
+end PresheafOfModules
+
+variable (R) in
+/-- This structure contains the data and axioms in order to
+produce a `PresheafOfModules R` from a collection of types
+equipped with module structures over the various rings `R.obj X`.
+(See the constructor `PresheafOfModules.mk'`.) -/
+structure CorePresheafOfModules where
+  /-- the datum of a type for each object in `Cᵒᵖ` -/
+  obj (X : Cᵒᵖ) : Type v
+  /-- the abelian group structure on the types `obj X` -/
+  addCommGroup (X : Cᵒᵖ) : AddCommGroup (obj X) := by infer_instance
+  /-- the module structure on the types `obj X` over the various rings `R.obj X` -/
+  module (X : Cᵒᵖ) : Module (R.obj X) (obj X) := by infer_instance
+  /-- the semi-linear restriction maps -/
+  map {X Y : Cᵒᵖ} (f : X ⟶ Y) : obj X →ₛₗ[R.map f] obj Y
+  /-- `map` is compatible with the identities -/
+  map_id (X : Cᵒᵖ) (x : obj X) : map (𝟙 X) x = x := by aesop_cat
+  /-- `map` is compatible with the composition -/
+  map_comp {X Y Z : Cᵒᵖ} (f : X ⟶ Y) (g : Y ⟶ Z) (x : obj X) :
+    map (f ≫ g) x = map g (map f x) := by aesop_cat
+
+-- this example is meant to test automation: the axioms for `CorePresheafOfModules` are
+-- automatically found if we use the data from `M : PresheafOfModules R`
+example (M : PresheafOfModules R) : CorePresheafOfModules R where
+  obj X := M.obj X
+  map f := M.map f
+
+namespace CorePresheafOfModules
+
+attribute [instance] addCommGroup module
+attribute [simp] map_id map_comp
+
+variable (M : CorePresheafOfModules R)
+
+/-- The presheaf of abelian groups attached to a `CorePresheafOfModules R`. -/
+@[simps]
+def presheaf : Cᵒᵖ ⥤ AddCommGroupCat.{v} where
+  obj X := AddCommGroupCat.of (M.obj X)
+  map f := AddCommGroupCat.ofHom (M.map f).toAddMonoidHom
+
+instance (X : Cᵒᵖ) : Module (R.obj X) (M.presheaf.obj X) := M.module X
+
+end CorePresheafOfModules
+
+namespace PresheafOfModules
+
+/-- Constructor for `PresheafOfModules R` based on a collection of types
+equipped with module structures over the various rings `R.obj X`, see
+the structure `CorePresheafOfModules`. -/
+def mk' (M : CorePresheafOfModules R) : PresheafOfModules R where
+  presheaf := M.presheaf
```

Comment:
Why is this in the `PresheafOfModules` namespace and not in the `CorePresheafOfModules` namespace? I would think the latter allows dot notation on `M`?

Reply:
I have changed this to `CorePresheafOfModules.toPresheafOfModules` and similar for the other.